docs/DevGuide/TensorExpressions.md

```diff
@@ -0,0 +1,411 @@
+\cond NEVER
+Distributed under the MIT License.
+See LICENSE.txt for details.
+\endcond
+# Writing tensor equations with TensorExpressions {#writing_tensorexpressions}
+
+\tableofcontents
+
+SpECTRE's `TensorExpression`s interface allows you to write tensor equations in
+SpECTRE in C++ with syntax that resembles tensor index notation. To use it,
+simply add this include to the top of your file:
+```
+#include "DataStructures/Tensor/Tensor.hpp"
+```
+The following guide assumes a basic understanding of the `Tensor` class and
+\ref tnsr "tnsr" type aliases.
+
+# Syntax {#te_syntax}
+`TensorExpression`s are arithmetic expressions of `Tensor`s that can be
+evaluated using `tenex::evaluate`. Terms used in the expression may be `Tensor`s
+or numbers (see [supported types](#te_data_type_support)).
+
+As a simple example of how `TensorExpression`s are used, if you would like to
+raise the index of some `Tensor` `R` with some spacetime metric `Tensor` `g`,
+i.e. \f$R^c{}_b = R_{ab} g^{ac}\f$, you can compute this with
+`TensorExpression`s by doing:
+```
+auto R_up = tenex::evaluate<ti::C, ti::b>(R(ti::a, ti::b) * g(ti::A, ti::C));
```

Comment:
What do you think about making these code examples snippets from a compiled cpp file? Then they're guaranteed to work. It means going through and actually moving them to a cpp file, so I leave it up to you to decide if it's worth the effort.

Reply:
Sure, I can do this! Would it be reasonable to make a new test file specifically dedicated to this guide? That way there is one spot that all of the demonstration examples are?

Reply:
Because right now, the tests are separated by operation

Reply:
Yea sounds good! The Brigand documentation does that too.

Reply:
We have example files for other things already that are just used to compile and be pretty docs things:

- `tests/Unit/Helpers/ParallelAlgorithms/Interpolation/Examples.hpp`
- `tests/Unit/Helpers/ControlSystem/Examples.hpp`